File: src/main/python/coolero/repositories/liquidctl_repo.py
#  Coolero - monitor and control your cooling and other devices
#  Copyright (c) 2021  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging.config
from typing import Optional, List, Dict, Tuple, Any, Union

import liquidctl
from liquidctl.driver.base import BaseDriver

from models.device_info import DeviceInfo
from models.device_status import DeviceStatus
from models.status import Status
from repositories.devices_repository import DevicesRepository
from services.device_extractor import DeviceExtractor

_LOG = logging.getLogger(__name__)


class LiquidctlRepo(DevicesRepository):
    """Repo for all Liquidctl devices"""

    _device_statuses: Dict[int, Tuple[DeviceStatus, BaseDriver]] = {}
    _device_info_extractor: DeviceExtractor

    def __init__(self) -> None:
        self._device_info_extractor = DeviceExtractor()
        super().__init__()
        _LOG.info('initialized with status: %s', self._device_statuses)

    @property
    def statuses(self) -> List[DeviceStatus]:
        return [device_status for device_status, _ in self._device_statuses.values()]

    def update_statuses(self) -> None:
        for device_status, lc_device in self._device_statuses.values():
            device_status.status = self._map_status(
                lc_device,
                lc_device.get_status()
            )
            _LOG.debug('Liquidctl device: %s status was updated with: %s',
                       device_status.device_name,
                       device_status.status)

    def shutdown(self) -> None:
        """Should be run on exit & shutdown, even in case of exception"""
        for _, lc_device in self._device_statuses.values():
            lc_device.disconnect()
        self._device_statuses.clear()
        _LOG.debug("Liquidctl Repo shutdown")

    def _initialize_devices(self) -> None:
        _LOG.debug("Initializing Liquidctl devices")
        # todo: try catches all over, like for when no connection is possible
        devices = liquidctl.find_liquidctl_devices()

        for index, device in enumerate(devices):
            device.connect()
            lc_init_status: List[Tuple] = device.initialize()
            _LOG.debug(f'Liquidctl device initialization response: {lc_init_status}')
            init_status = self._map_status(device, lc_init_status)
            device_info = self._extract_device_info(device)
            device_status = DeviceStatus(
                _device_name=device.description,
                _status_current=init_status,
                _lc_device_id=index,
                _lc_driver_type=type(device),
                _lc_init_firmware_version=init_status.firmware_version,
                _device_info=device_info
            )
            # get the status after initialization to fill with complete data right away
            device_status.status = self._map_status(device, device.get_status())
            self._device_statuses[index] = (device_status, device)

    def _map_status(self, device: BaseDriver, lc_status: List[Tuple]) -> Status:
        status_dict = self._convert_status_to_dict(lc_status)
        return self._device_info_extractor.extract_status_from(device, status_dict)

    @staticmethod
    def _convert_status_to_dict(lc_status: List[Tuple]) -> Dict[str, Union[str, int, float]]:
        return dict((str(lc_property).strip().lower(), value) for lc_property, value, unit in lc_status)

    def _extract_device_info(self, device: BaseDriver) -> Optional[DeviceInfo]:
        return self._device_info_extractor.extract_info_from(device)

File: coolercontrol/src/liquidctl/liquidctl_repo.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2022  Guy Boldon
 * |
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * |
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * |
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 ******************************************************************************/


use std::borrow::{Borrow};
use std::clone::Clone;
use std::collections::HashMap;
use std::ops::{Deref, DerefMut};
use std::str::FromStr;
use std::string::ToString;
use std::time::{Duration, Instant};

use anyhow::{anyhow, bail, Result};
use async_trait::async_trait;
use const_format::concatcp;
use log::{debug, error, info, warn};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use tokio::time::sleep;
use zbus::export::futures_util::future;

use crate::Device;
use crate::device::{DeviceType, Status};
use crate::liquidctl::base_driver::BaseDriver;
use crate::liquidctl::device_mapper::DeviceMapper;
use crate::repository::Repository;
use crate::setting::Setting;

pub struct LiquidctlRepo {
    client: Client,
    device_mapper: DeviceMapper,
    devices: RwLock<Vec<Device>>,
}

type LCStatus = Vec<(String, String, String)>;

const LIQCTLD_ADDRESS: &str = "http://127.0.0.1:11986";
const LIQCTLD_HANDSHAKE: &str = concatcp!(LIQCTLD_ADDRESS, "/handshake");
const LIQCTLD_DEVICES: &str = concatcp!(LIQCTLD_ADDRESS, "/devices");
const LIQCTLD_DEVICES_CONNECT: &str = concatcp!(LIQCTLD_ADDRESS, "/devices/connect");
const LIQCTLD_LEGACY690: &str = concatcp!(LIQCTLD_ADDRESS, "/devices/{}/legacy690");
const LIQCTLD_INITIALIZE: &str = concatcp!(LIQCTLD_ADDRESS, "/devices/{}/initialize");
const LIQCTLD_STATUS: &str = concatcp!(LIQCTLD_ADDRESS, "/devices/{}/status");
const LIQCTLD_QUIT: &str = concatcp!(LIQCTLD_ADDRESS, "/quit");


impl LiquidctlRepo {
    pub async fn new() -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(10))
            .build()?;
        // todo: self generated certs
        LiquidctlRepo::establish_connection(&client).await?;
        info!("Communication established with Liqctld.");
        Ok(LiquidctlRepo {
            client,
            device_mapper: DeviceMapper::new(),
            devices: RwLock::new(vec![]),
        })
    }

    async fn establish_connection(client: &Client) -> Result<()> {
        let mut retry_count: u8 = 0;
        while retry_count < 5 {
            match client.get(LIQCTLD_HANDSHAKE).send().await {
                Ok(response) =>
                    return match response.json::<HandshakeResponse>().await {
                        Ok(handshake_response) => if handshake_response.shake {
                            Ok(())
                        } else {
                            Err(anyhow!(
                                    "Incorrect Handshake confirmation. Shake: {}",
                                    handshake_response.shake)
                            )
                        }
                        Err(err) => Err(anyhow!(err))
                    },
                Err(err) =>
                    error!(
                    "Could not establish communication with liqctld socket connection, retry #{}. \n{}",
                    retry_count + 1, err
                )
            };
            sleep(Duration::from_secs(1)).await;
            retry_count += 1;
        }
        bail!("Failed to connect to liqctld after {} tries", retry_count);
    }

    pub async fn get_devices(&mut self) -> Result<()> {
        let devices_response = self.client.get(LIQCTLD_DEVICES)
            .send().await?
            .json::<DevicesResponse>().await?;
        for device_response in devices_response.devices {
            let device_type = match self.map_device_type(&device_response) {
                None => {
                    warn!("Device is currently not supported: {:?}", device_response.device_type);
                    continue;
                }
                Some(d_type) => d_type
            };
            // self.devices.borrow_mut().push(
            self.devices.get_mut().push(
                Device {
                    name: device_response.description,
                    d_type: DeviceType::Liquidctl,
                    type_id: device_response.id,
                    status_history: vec![],
                    lc_driver_type: Some(device_type),
                    lc_init_firmware_version: None,
                    info: None,
                }
            );
        }
        debug!("List of received Devices: {:?}", self.devices);
        Ok(())
    }

    pub async fn connect_devices(&self) -> Result<()> {
        let connection_response = self.client.post(LIQCTLD_DEVICES_CONNECT)
            .send().await?
            .json::<ConnectionResponse>().await?;
        if connection_response.connected {
            info!("All Liquidctl Devices connected");
            Ok(())
        } else {
            Err(anyhow!("Incorrect Connect Devices Response: {}", connection_response.connected))
        }
    }

    fn map_device_type(&self, device: &DeviceResponse) -> Option<BaseDriver> {
        BaseDriver::from_str(device.device_type.as_str())
            .ok()
            .filter(|driver| self.device_mapper.is_device_supported(driver))
    }

    fn map_status(&self,
                  device_type: &BaseDriver,
                  lc_statuses: &LCStatus,
                  device_id: &u8,
    ) -> Status {
        let mut status_map: HashMap<String, String> = HashMap::new();
        for lc_status in lc_statuses {
            status_map.insert(lc_status.0.to_lowercase(), lc_status.1.clone());
        }
        self.device_mapper.extract_status(device_type, &status_map, device_id)
    }

    async fn call_initialize_concurrently(&self) {
        let mut devices_borrowed = self.devices.write().await;
        let mut futures = vec![];
        for device in devices_borrowed.deref_mut() {
            futures.push(self.call_initialize_per_device(device));
        }
        let results: Vec<Result<()>> = future::join_all(futures).await;
        for result in results {
            match result {
                Ok(_) => {}
                Err(err) => error!("Error getting initializing device: {}", err)
            }
        }
    }

    async fn call_initialize_per_device(&self, device: &mut Device) -> Result<()> {
        let status_response = self.client.borrow()
            .post(LIQCTLD_INITIALIZE.replace("{}", device.type_id.to_string().as_str()))
            .json(&InitializeRequest { pump_mode: None })
            .send().await?
            .json::<StatusResponse>().await?;
        let init_status = self.map_status(
            device.lc_driver_type.as_ref().unwrap(),
            &status_response.status,
            &device.type_id,
        );
        device.lc_init_firmware_version = init_status.firmware_version.clone();
        device.set_status(init_status);
        Ok(())
    }

    async fn call_status_concurrently(&self) {
        let mut devices_borrowed = self.devices.write().await;
        let mut futures = vec![];
        for device in devices_borrowed.deref_mut() {
            futures.push(self.call_status_per_device(device));
        }
        let results: Vec<Result<()>> = future::join_all(futures).await;
        for result in results {
            match result {
                Ok(_) => {}
                Err(err) => error!("Error getting status from device: {}", err)
            }
        }
    }

    async fn call_status_per_device(&self, device: &mut Device) -> Result<()> {
        let status_response = self.client.borrow()
            .get(LIQCTLD_STATUS.replace("{}", device.type_id.to_string().as_str()))
            .send().await?  // todo: will .with_context help here to determine the device it has issues with?
            .json::<StatusResponse>().await?;
        let init_status = self.map_status(
            device.lc_driver_type.as_ref().unwrap(),
            &status_response.status,
            &device.type_id,
        );
        debug!("Status updated for LC Device #{} with: {:?}", &device.type_id, &init_status);
        device.set_status(init_status);
        Ok(())
    }
}

#[async_trait]
impl Repository for LiquidctlRepo {
    async fn initialize_devices(&self) -> Result<()> {
        debug!("Starting Device Initialization");
        let start_initialization = Instant::now();
        self.call_initialize_concurrently().await;
        debug!("Initialized Devices: {:?}", self.devices.read().await);
        debug!(
            "Time taken to initialize all liquidctl devices: {:?}", start_initialization.elapsed()
        );
        info!("All liquidctl devices initialized");
        Ok(())
    }

    async fn devices(&self) -> Vec<Device> {
        let mut vec = vec![];
        for dev in self.devices.read().await.deref() {
            vec.push(dev.clone())  // Currently clones all devices
        }
        vec
    }

    async fn update_statuses(&self) -> Result<()> {
        debug!("Updating all Liquidctl device statuses");
        let start_initialization = Instant::now();
        self.call_status_concurrently().await;
        debug!(
            "Time taken to get status for all liquidctl devices: {:?}",
            start_initialization.elapsed()
        );
        info!("All liquidctl device statuses updated");
        Ok(())
    }

    async fn shutdown(&self) -> Result<()> {
        debug!("Shutting down Liquidctl Repo");
        let quit_response = self.client
            .post(LIQCTLD_QUIT)
            .send().await?
            .json::<QuitResponse>().await?;
        return if quit_response.quit {
            info!("Quit successfully sent to Liqctld");
            Ok(())
        } else {
            Err(anyhow!("Incorrect quit response from liqctld: {}", quit_response.quit))
        };
    }

    async fn apply_setting(&self, device_type_id: u8, setting: Setting) -> Result<()> {
        todo!()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct HandshakeResponse {
    shake: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct QuitResponse {
    quit: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct DeviceResponse {
    id: u8,
    description: String,
    device_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct DevicesResponse {
    devices: Vec<DeviceResponse>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ConnectionResponse {
    connected: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct InitializeRequest {
    pump_mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct StatusResponse {
    status: LCStatus,
}

